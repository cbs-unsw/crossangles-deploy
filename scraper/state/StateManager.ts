import AWS from 'aws-sdk';

export class StateManager {
  private readonly tableName = 'ScraperData';

  constructor (
    private readonly client: AWS.DynamoDB.DocumentClient,
  ) {}

  async init (ddb: AWS.DynamoDB) {
    try {
      await ddb.createTable({
        TableName: this.tableName,
        KeySchema: [
          {
            AttributeName: 'campus',
            KeyType: 'HASH',
          },
          {
            AttributeName: 'key',
            KeyType: 'RANGE',
          },
        ],
        AttributeDefinitions: [
          {
            AttributeName: 'campus',
            AttributeType: 'S',
          },
          {
            AttributeName: 'key',
            AttributeType: 'S',
          },
        ],
        BillingMode: 'PAY_PER_REQUEST',
      }).promise();
    } catch (err) {
      if (err.code !== 'ResourceInUseException') {
        throw err;
      }
    }
  }

  async drop (ddb: AWS.DynamoDB) {
    try {
      await ddb.deleteTable({
        TableName: this.tableName,
      }).promise();
    } catch (err) {
      if (err.code !== 'ResourceNotFoundException') {
        throw err;
      }
    }
  }

  async get (campus: string, key: string) {
    try {
      const response = await this.client.get({
        TableName: this.tableName,
        Key: { campus, key },
      }).promise();
      const result = response.Item.value;
      return result;
    } catch (err) {
      return undefined;
    }
  }

  async set (campus: string, key: string, value: any) {
    await this.client.put({
      TableName: this.tableName,
      Item: {
        campus,
        key,
        value,
      },
    }).promise();
  }

  async delete (campus: string, key: string) {
    await this.client.delete({
      TableName: this.tableName,
      Key: {
        campus,
        key,
      },
    }).promise();
  }
}

export default StateManager;
