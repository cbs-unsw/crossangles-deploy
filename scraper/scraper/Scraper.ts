import axios from 'axios';
import cheerio from 'cheerio';
import axiosRetry from 'axios-retry';
import AsyncQueue from './AsyncQueue';
import HTMLCache from './HTMLCache';
import { CourseData } from '../../app/src/state/Course';
import { Meta } from '../../app/src/state/Meta';

axiosRetry(axios, { retryDelay: axiosRetry.exponentialDelay });

export interface CampusData {
  campus: string,
  courses: CourseData[],
  meta: Meta,
  current: boolean,
}

export class Scraper {
  maxRequests: number = 10;
  cache: HTMLCache;

  constructor (cache?: HTMLCache) {
    this.cache = cache || new HTMLCache();
  }

  async scrapePages<T> (urls: string[], handler: (page: CheerioStatic, url: string) => Promise<T>) {
    const queue = new AsyncQueue<string, T>(this.maxRequests);
    queue.enqueue(urls);
    const processor = async (url: string) => {
      const content = await this.getPageContent(url).catch(error => {
        console.error(`Error while fetching "${url}":`, error);
        return null;
      });
      if (content === null) return null;

      return handler(content, url).catch(error => {
        console.error(`Error while scraping "${url}":`, error);
        return null;
      });
    };
    const parsingPromises = await queue.run(processor);
    return await Promise.all(parsingPromises);
  }

  async getPageContent (url: string) {
    let content: string;
    let loadedFromCache = false;
    if (this.cache.has(url)) {
      content = this.cache.get(url)!;
      loadedFromCache = true;
    } else {
      const response = await axios.get<string>(url);
      content = response.data;
    }

    const page = cheerio.load(content);

    if (!loadedFromCache) {
      this.cache.set(url, page.html());
    }

    return page;
  }
}
