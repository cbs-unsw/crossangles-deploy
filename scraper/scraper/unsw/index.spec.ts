import { scrapeUNSW } from './index';
import { Scraper } from '../Scraper';
import StateManager from '../../state/StateManager';

describe('scrapeUNSW', () => {
  const cacheFile = 'unsw-snapshot.json.br';
  const scraper = new Scraper();
  const cache = scraper.cache;
  const cachePromise = cache.load(cacheFile).catch(() => {});
  const state: StateManager = { get: jest.fn(), set: jest.fn(), getBlob: jest.fn(), setBlob: jest.fn() } as any;

  it('gives consistent output', async () => {
    await cachePromise;
    const result = await scrapeUNSW({ scraper, state, forceUpdate: true });
    expect(result).not.toBeNull();
    const courses = result![0].courses;
    for (let i = 0; i < courses.length; ++i) {
      if (courses[i].isAdditional) {
        courses[i] = {
          ...courses[i],
          metadata: undefined,
          streams: [],
          description: undefined,
          defaultColour: undefined,
        };
      }
    }
    expect(courses.filter(c => c.isAdditional)).toMatchSnapshot();
    await cache.write(cacheFile);
  })
})
