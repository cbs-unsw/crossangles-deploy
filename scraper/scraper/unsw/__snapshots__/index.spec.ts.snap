// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`scrapeUNSW gives consistent output 1`] = `
Array [
  Object {
    "autoSelect": true,
    "code": "CBS",
    "defaultColour": undefined,
    "description": undefined,
    "isAdditional": true,
    "metadata": undefined,
    "name": "Campus Bible Study",
    "streams": Array [],
  },
]
`;
